### PR TITLE
Check for duplicated joint names in a trajectory

### LINKED ISSUE #28: Improve JTA goal validation to also check for duplicate joint names
As per subject.

The current implementation of the `FollowJointTrajectory` action goal validation does not check whether incoming `JointTrajectory`s have duplicated joint names in the `joint_names` field.

This could allow malformed trajectories to be considered OK to execute, while they should be rejected.

### PR DESCRIPTION
**Pull request by ted-miller**
Fixes #28 
Not yet tested due to https://github.com/Yaskawa-Global/micro_ros_motoplus/issues/132

### CODE REVIEW

**Comment by ted-miller** [comment]
Tested and working

**Comment by gavanderhoorn** [comment]
`msbuild` CI is failing due to something I introduced in #158.

I'll add a commit to this PR to fix it.

**Comment by gavanderhoorn** [comment]
Actually: could you try rebasing on `main` @ted-miller?

**Comment by ted-miller** [comment]
ignore that.

**Comment by ted-miller** [comment]
Ok. Now I think it'll work

**Comment by gavanderhoorn** [comment]
I've rebased this branch to include the fix from #164.

**Comment by gavanderhoorn** [comment]
I can't test this right now, but you have.

@ted-miller: could you maybe describe how you tested it?

**Comment by ted-miller** [comment]
Using my FJT script, I duplicated the first axis joint name on the second axis.
When I run the script, the robot returns the 214 error code with the error message.
The robot did not attempt to move.

File: src/MotionControl.c [hunk 1]
```diff
@@ -94,6 +94,17 @@ Init_Trajectory_Status Ros_MotionControl_Init(rosidl_runtime_c__String__Sequence
         int  jointIndexInCtrlGroup;
         CtrlGroup* ctrlGroup;
 
+        //check to ensure there are no duplicate joint names in the list
+        for (checkForDupIndex = jointIndexInTraj; checkForDupIndex < sequenceGoalJointNames->size; checkForDupIndex += 1)
```
Comment (gavanderhoorn) [inline src/MotionControl.c:98]: It's late here, but wouldn't this lead to `checkForDupIndex == jointIndexInTraj` on the first iteration? At the very beginning of the `jointIndexInTraj` loop, that'd give `checkForDupIndex == jointIndexInTraj == 0`, which would compare the first joint name with itself?

File: src/MotionControl.c [hunk 2]
```diff
@@ -94,6 +94,17 @@ Init_Trajectory_Status Ros_MotionControl_Init(rosidl_runtime_c__String__Sequence
...
+        {
+            if (strncmp(sequenceGoalJointNames->data[jointIndexInTraj].data,
+                sequenceGoalJointNames->data[checkForDupIndex].data,
+                MAX_JOINT_NAME_LENGTH) == 0)
+            {
+                Ros_Debug_BroadcastMsg("Joint name [%s] is used for multiple joints in the trajectory.", sequenceGoalJointNames->data[jointIndexInTraj].data);
```
Comment (gavanderhoorn) [inline src/MotionControl.c:104]:
```suggestion
                Ros_Debug_BroadcastMsg("Joint name [%s] is used for multiple joints in the trajectory (indices: %d and %d).", sequenceGoalJointNames->data[jointIndexInTraj].data, jointIndexInTraj, checkForDupIndex);
```

File: src/MotionControl.c (hunk 1 above)
Comment (ted-miller) [reply src/MotionControl.c:98]: Yeah, that's supposed to be `+1`. Good catch.

**Comment by gavanderhoorn** [comment]
> Using my FJT script, I duplicated the first axis joint name on the second axis. When I run the script, the robot returns the 214 error code with the error message. The robot did not attempt to move.

might want to check also with a joint at index `!= 0`.

**Review by gavanderhoorn — APPROVED** [review]
Updated version looks OK.

If you could test with repeated joint names at indices other than 0, this can be merged.

**Comment by ted-miller** [comment]
Tested again:

'group_1/joint_1',
'group_1/joint_2',
'group_1/joint_3',
'group_1/joint_4',
'group_1/joint_5',
'group_1/joint_6',

'group_2/joint_1',
'group_2/joint_2',
**'group_1/joint_3'**,
'group_2/joint_4',
'group_2/joint_5',
'group_2/joint_6',

'group_3/joint_1'

Client:
`control_msgs.action.FollowJointTrajectory_Result(error_code=-100214, error_string='The trajectory contains duplicate joint names.')`

Debug Log:
```
[1696258937.30257201] [192.168.1.32:58802]: 2023-10-02 15:02:17.301923 Trajectory contains 61 points
[1696258937.30277658] [192.168.1.32:58802]: 2023-10-02 15:02:17.302123 Joint name [group_1/joint_3] is used for multiple joints in the trajectory (indices: 2 and 8).
[1696258937.30299282] [192.168.1.32:58802]: 2023-10-02 15:02:17.302323 FollowJointTrajectory - Goal request rejected
[1696258937.30308771] [192.168.1.32:58802]: 2023-10-02 15:02:17.302323 The trajectory will be accepted and then immediately aborted
[1696258937.30314612] [192.168.1.32:58802]: 2023-10-02 15:02:17.302323 The trajectory contains duplicate joint names.

```

**Comment by gavanderhoorn** [comment]
off-topic, but: what's up with the output of your debug logger?

It's not supposed to look like that any more since #129 and some older PRs.

**Comment by ted-miller** [comment]
I've got an old copy of the script in my home directory for convenience. I just haven't updated on my ubuntu machine.

**Comment by gavanderhoorn** [comment]
You could use a symlink and place it in `$HOME/bin`. That would avoid the copy and always keep it up-to-date.

**Pull request merged**